ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+    }
+    
+    @Test
+    public void initialisationTest() {
+        assertTrue(testCarpark instanceof Carpark);
+        assertTrue(testIS instanceof CarSensor);
+        assertTrue(testOS instanceof CarSensor);
+        assertTrue(testExit instanceof ExitController);
+    }
+    
+    @Test
+    public void testNormalFlowSeason() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testSeason.recordUsage(mockRecord);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");
+    }
+    
+    @Test
+    public void testNormalFlowAdhoc() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");            //checks state transition and message display
+        
+        when(mockAdhocDAO.findTicketByBarcode("Atest")).thenReturn(testAdhoc);
+        testAdhoc.pay(1234, 1234);
+        testExit.ticketInserted(testAdhoc.getBarcode());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");
+    }
+    
+    @Test
+    public void testAlternateFlowAdhoc() {          //test accurate detection of rejected adhoc ticket
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        assertEquals(testExit.getState(), "WAITING");
+        
+        when(mockAdhocDAO.findTicketByBarcode("Atest")).thenReturn(null);

```

Comment:
you cannot do this, you are not 'testing accurate detection of rejected adhoc ticket' because you are telling it what valus to return

Reply:
As above, adhoc testing was part of Aaron's stuff, I'm just simulating the situation in which the ticket isn't found. The adhocTicketDAO.findTicketByBarcode() method is called by Carpark and it's value returned there, so the proper validation is still done.

Reply:
use comments for readability
would recommend putting setup stuff at the top of your methods